athenz-identity-provider-service/src/test/java/com/yahoo/vespa/hosted/athenz/instanceproviderservice/ca/CertificateSignerTest.java

```diff
@@ -0,0 +1,134 @@
+package com.yahoo.vespa.hosted.athenz.instanceproviderservice.ca;
+
+import com.yahoo.test.ManualClock;
+import org.bouncycastle.asn1.pkcs.PKCSObjectIdentifiers;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x509.Extension;
+import org.bouncycastle.asn1.x509.Extensions;
+import org.bouncycastle.asn1.x509.ExtensionsGenerator;
+import org.bouncycastle.asn1.x509.GeneralName;
+import org.bouncycastle.asn1.x509.GeneralNames;
+import org.bouncycastle.operator.ContentSigner;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+import org.bouncycastle.pkcs.PKCS10CertificationRequest;
+import org.bouncycastle.pkcs.PKCS10CertificationRequestBuilder;
+import org.bouncycastle.pkcs.jcajce.JcaPKCS10CertificationRequestBuilder;
+import org.junit.Test;
+
+import java.math.BigInteger;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/**
+ * @author freva
+ */
+public class CertificateSignerTest {
```

Comment:
Consider writing unit tests that tests the code through the public interface. It makes the test more robust and also easier to refactor the implementation later on without resorting to updating all unit tests.

Reply:
I feel like its usually the opposite? Testing a public method that relies on some internal methods often means that it's harder to set up the test, more variables are needed. It's also harder to see what exactly failed in cases where the method returns/throws the same thing.

In this case both `verify*` methods throw `IllegalArgumentException`, so you would have to check the message to see which one of them complained.
If we were to make one of the `verify`s stricter, then there is a good chance all the test would fail because usually the test case is set up as simply as possible.

Reply:
I agree that it's harder to write unit tests in that way, and retro-fitting such tests to existing code could be too cumbersome. Imho it is most times worth the effort and ensures that the class is tested the way it is supposed to be used.